### DISCUSSION TITLE
[@types/esquery] problems

### DISCUSSION THREAD

**Discussion by jakerdy** [DefinitelyTyped/DefinitelyTyped#69575] [Issues with a @types package] [answered]
Hi, I found couple issues with types for esquery package...

## Wrong signature of `traverse` method

Query should be parsed before invoking `traverse`, otherwise it will thorw `Unknown selector type: undefined`.

```typescript
const ast = parseScript(...);
function handler(n, p){ ... }

// will thorw
esquery.traverse(ast, `[type="Identifier"]`, handler)

// works, but typescript errors because of wrong type
const sel = esquery.parse(`[type="Identifier"]`);
esquery.traverse(ast, sel, handler)

```

## Named imports not working

I use esquery in browser with vite.

When esquery apis imported via named imports:
```typescript
import { query, parse, traverse } from "esquery";
```

Browser says: `Uncaught SyntaxError: The requested module '/node_modules/.vite/deps/esquery.js?v=c82eb67a' does not provide an export named 'parse' (at formula.ts:3:10)`

But, if I import via default import like this:

```typescript
import esquery from "esquery";
```

Errors disapear. I don't know if this issue related to typings, or vite, but it is here...

## Environment

- vite: 5.2.10
- esprima: 4.0.1
- esquery: 1.5.0
- @types/esprima: 4.0.6
- @types/esquery: 1.5.3

## Ping

@cherryblossom000
@bradzacher
@brettz9

[a comment by an automated account is not shown]

**Comment by brettz9** [comment]
esquery does not provide named imports, so I think that aspect is working as intended.

**Comment by cherryblossom000** [comment, marked as answer]
Thanks for pointing this out! I’ve submitted a PR to fix the types of the `traverse` function: #69579
